src/operator/join.ts

```diff
@@ -0,0 +1,172 @@
+              private resultSelector: (left: T1, right: T2) => R) {
+  }
+
+  call(subscriber: Subscriber<R>): Subscriber<any> {
+    return new JoinSubscriber(subscriber, this.leftDurationSelector, this.rightDurationSelector, this.resultSelector);
+  }
+}
+
+class JoinSubscriber<T1, T2, D1, D2, R> extends OuterSubscriber<any, R> {
+  private observables: Observable<T1 | T2>[] = [];
+  private leftId: number = 2;
+  private rightId: number = 3;
+  private leftMap: Map<number, T1> = new Map<number, T1>();
+  private rightMap: Map<number, T2> = new Map<number, T2>();
+  private leftDone: boolean = false;
+  private rightDone: boolean = false;
+
+  constructor(destination: Subscriber<R>,
+              private leftDurationSelector: (value: T1) => Observable<D1>,
+              private rightDurationSelector: (value: T2) => Observable<D2>,
+              private resultSelector: (left: T1, right: T2) => R) {
+    super(destination);
+  }
+
+  protected _next(value: Observable<T1 | T2>): void {
+    this.observables.push(value);
+  }
+
+  protected _complete(): void {
+    const os = this.observables;
+    this.add(subscribeToResult(this, os[0], null, 0));
+    this.add(subscribeToResult(this, os[1], null, 1));
+  }
+
+  notifyNext(outerValue: any, innerValue: any,
+             outerIndex: number, innerIndex: number,
+             innerSub: InnerSubscriber<any, R>): void {
+    if (outerIndex === 0) {
+      this.leftNext(innerValue);
+    } else if (outerIndex === 1) {
+      this.rightNext(innerValue);
+    }
+  }
+
+  private leftNext(value: T1): void {
+    const id = this.leftId;
+    this.leftId += 2;
+    this.leftMap.set(id, value);
+    const success = this.tryDurationSelector(this.leftDurationSelector, value, id);
+    if (success) {
+      this.rightMap.forEach((v: T2) => {
+        this.tryResultSelector(value, v);
+      });
+    }
+  }
+
+  private rightNext(value: T2): void {
+    const id = this.rightId;
+    this.rightId += 2;
+    this.rightMap.set(id, value);
+    const success = this.tryDurationSelector(this.rightDurationSelector, value, id);
+    if (success) {
+      this.leftMap.forEach((v: T1) => {
+        this.tryResultSelector(v, value);
+      });
+    }
+  }
+
+  private tryDurationSelector(durationSelector: (value: T1 | T2) => Observable<D1 | D2>,
+                              value: T1 | T2, id: number): boolean {
+    let duration: Observable<D1 | D2>;
+    try {
+      duration = durationSelector(value);
+    } catch (err) {
+      this.destination.error(err);
+      return false;
+    }
+    const durationObservable = take.call(duration, 1);
+    this.add(subscribeToResult(this, durationObservable, null, id));
+    return true;
+  }
+
+  private tryResultSelector(leftValue: T1, rightValue: T2): void {
+    let result: R;
+    try {
+      result = this.resultSelector(leftValue, rightValue);
+    } catch (err) {
+      this.destination.error(err);
+      return;
+    }
+    this.destination.next(result);
```

Comment:
Sorry, just getting back to this... it's a bit of both. 

1. Every call to `this.something` incurs a property/hashtable type lookup. If you stuff it in a variable, you don't have to make that lookup. It's not a _huge_ optimization, but it's a simple one.

2. When minified, it'll be smaller:

`this.foo.something(); this.foo.whatever();` minified is the same thing -> `this.foo.something();this.foo.whatever()`.

But with the const before hand:
`const foo = this.foo; foo.something(); foo.whatever();` -> `const a = this.foo; a.something(); a.whatever()`